src/shogun/mathematics/linalg/internal/implementation/Mean.h

```diff
@@ -0,0 +1,984 @@
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation are those
+ * of the authors and should not be interpreted as representing official policies,
+ * either expressed or implied, of the Shogun Development Team.
+ */
+
+#ifndef MEAN_IMPL_H_
+#define MEAN_IMPL_H_
+
+#include <shogun/lib/config.h>
+#include <shogun/lib/SGMatrix.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/io/SGIO.h>
+#include <shogun/mathematics/linalg/internal/Block.h>
+#include <shogun/mathematics/linalg/internal/implementation/Sum.h>
+
+#include <shogun/mathematics/eigen3.h>
+
+#ifdef HAVE_VIENNACL
+#include <shogun/mathematics/linalg/internal/implementation/Scale.h>
+#include <shogun/mathematics/linalg/internal/opencl_util.h>
+#include <shogun/lib/GPUMatrix.h>
+#include <shogun/lib/GPUVector.h>
+
+#endif
+
+#include <string>
+
+namespace shogun
+{
+
+namespace linalg
+{
+
+/**
+ * All backend specific implementations are defined within this namespace
+ */
+namespace implementation
+{
+
+/**
+ * @brief Generic class mean which provides a static compute method. This class
+ * is specialized for different types of matrices and backend, providing a means
+ * to deal with various matrices directly without having to convert
+ */
+template <enum Backend,class Matrix>
+struct mean
+{
+	/** Scalar type */
+	typedef typename Matrix::Scalar T;
+
+	/**
+	 * Method that computes the mean of a dense matrix
+	 *
+	 * @param m the matrix whose mean has to be computed
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @return the mean computed as \f$\mean_{i,j}m_{i,j}\f$
+	 */
+	static T compute(Matrix m, bool no_diag);
+
+	/**
+	 * Method that computes the mean of dense matrix blocks
+	 *
+	 * @param b the matrix-block whose mean has to be computed
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @return the mean computed as \f$\mean_{i,j}b_{i,j}\f$
+	 */
+	static T compute(Block<Matrix> b, bool no_diag);
+};
+
+/**
+ * @brief Generic class mean symmetric which provides a static compute method. This class
+ * is specialized for different types of matrices and backend, providing a means
+ * to deal with various matrices directly without having to convert
+ */
+template <enum Backend,class Matrix>
+struct mean_symmetric
+{
+	/** Scalar type */
+	typedef typename Matrix::Scalar T;
+
+	/**
+	 * Method that computes the mean of a symmetric dense matrix
+	 *
+	 * @param m the matrix whose mean has to be computed
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @return the mean computed as \f$\mean_{i,j}m_{i,j}\f$
```

Comment:
Oops..Sorry I had no experience with Latex and I just did a simple replacement.
Will work on the code. Thanks!